src/main/java/de/tum/in/www1/artemis/repository/ProgrammingExerciseRepository.java

```diff
@@ -121,6 +121,19 @@
     @Query("select pe from ProgrammingExercise pe where pe.assessmentType <> 'AUTOMATIC' and pe.dueDate > :#{#dateTime}")
     List<ProgrammingExercise> findAllByManualAssessmentAndDueDateAfterDate(@Param("dateTime") ZonedDateTime dateTime);
 
+    /**
+     * Returns all programming exercises that have a due date after {@code dateTime} and have tests marked with
+     * {@link de.tum.in.www1.artemis.domain.enumeration.Visibility#AFTER_DUE_DATE} but no buildAndTestStudentSubmissionsAfterDueDate.
+     * @param dateTime the time after which the due date of the exercise has to be
+     * @return List<ProgrammingExercise> (can be empty)
+     */
+    @Query("""
+                select pe from ProgrammingExercise pe
+                where pe.dueDate > :#{#dateTime} and pe.buildAndTestStudentSubmissionsAfterDueDate is null
+                    and exists (select tc from ProgrammingExerciseTestCase tc where tc.exercise = pe and tc.visibility = 'AFTER_DUE_DATE')
+            """)
+    List<ProgrammingExercise> findAllByDueDateAfterDateWithTestsAfterDueDateWithoutBuildStudentSubmissionsDate(@Param("dateTime") ZonedDateTime dateTime);
```

Comment:
That method name is quite the mouthful. is it still reasonably easy to understand what is meant? Do you have a suggestion for a better name?